File: outco/getDuplicates.js
/**
 *  Get Duplicates
 *
 *  Given an array of values, return only the values that have duplicates in the
 *  array
 *
 *  Parameters
 *  Input: arr {Array}
 *  Output: {Array}
 *
 *  Constraints
 *  Time: O(N)
 *  Space: O(N)
 *
 *  Examples
 *  [1, 2, 4, 2] --> [2]
 *  [3, 2, 3, 2, 3, 3, 4] --> [3, 2]
 *  [1, 2, 3, 4] --> []
 */

// initialize a hash map
// loop over the array
// each element in the array is a key
// if an element was already observed increment the element's counter
// init duplicatesArray
// for each key in the hash map
// append keys from key value pairs with values greater than 1

const getDuplicates = (arr) => {
    let hash = {};
    for ( let i = 0; i<arr.length; i++){
      if (!hash[arr[i]]) {
        hash[arr[i]] = 1
      } else {
        hash[arr[i]]++
      }
    }
    let duplicatesArray = [];
    for (let key in hash){
      if (hash[key] > 1){
        duplicatesArray.push(parseInt(key));
      } 
    }
    return console.log(duplicatesArray)
}

getDuplicates([1, 2, 4, 2]) // [2]
getDuplicates([3, 2, 3, 2, 3, 3, 4]) //--> [3, 2]
getDuplicates([1, 2, 3, 4]) // --> []

File: c++journey/dsaDomains/frequencyCount/getDuplicates.cpp
/**
*  Get Duplicates
*  
*  Given an array of values, return only the values that have duplicates in the
*  array
*
*  Parameters
*  Input: arr {int array}
*  Output: {List of ints}
*
*  Constraints
*  Time: O(N)
*  Space: O(N)
*
*  Examples
*  [1, 2, 4, 2] --> [2]
*  [3, 2, 3, 2, 3, 3, 4] --> [3, 2]
*  [1, 2, 3, 4] --> []
*/
#include <array>
#include <cctype>
#include <cmath>
#include <iterator>
#include <ostream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <iostream>
#include <array>
#include <sstream>
#include <cctype>
using namespace std;

/* 
* 1. Initialize frequencyMap<int, int> and outputVec<int, int>.
* 2. Loop over input vector to build up the frequency map.
* 3. Loop through the frequencyMap.
*       --> if the key's frequency is 2 or higher,
*       --> append the key to the outputVec.
* 4. Return the outputVec.
 */
vector<int> getDuplicates(vector<int> vecX){
    vector<int> outputVec;
    unordered_map<int, int> frequencyMap;
    for (const int intX : vecX){
        if (frequencyMap.find(intX) != frequencyMap.end()){
            frequencyMap[intX] = frequencyMap[intX] + 1;
        }
        if (frequencyMap.find(intX) == frequencyMap.end()){
            frequencyMap[intX] = 1;
        }
    }
    for (const auto keyValue : frequencyMap){
        if (keyValue.second >= 2){
            outputVec.push_back(keyValue.first);
        }
    }
    return outputVec;
}

void printVector(const vector<int>& vec){
    for (int valueInt : vec){
        cout << valueInt << " ";
    }
    cout << endl;
}

void test(){
    vector<int> resultVec1 = getDuplicates({1, 2, 4, 2}); // {2}
    vector<int> resultVec2 = getDuplicates({2, 2, 3, 3, 3, 3, 4}); // {2, 3}
    vector<int> resultVec3 = getDuplicates({1, 2, 3, 4}); // {}
    vector<int> resultVec4 = getDuplicates({1, 1, 5, 5, 8, 8}); // {1, 5, 8}
    cout << "Result 1: " << endl;
    printVector(resultVec1);
    cout << "Result 2: " << endl;
    printVector(resultVec2);
    cout << "Result 3: " << endl;
    printVector(resultVec3);
    cout << "Result 4: " << endl;
    printVector(resultVec4);
}   

int main(){
    test();
    return 0;
}
/* Test results
Result 1: 
2 
Result 2: 
3 2 
Result 3: 

Result 4: 
8 5 1  */